[JIT] Enable ModuleList non-literal indexing

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53410 [JIT] Enable ModuleList non-literal indexing**

**Summary**
This commit enables indexing into `ModuleList` using a non-literal
index if the LHS of the assignment statement of which the indexing is
the RHS is annotated with an interface type.

This feature already exists for `ModuleDict`, and this commit builds on
top of that implementation. A `prim::ModuleContainerIndex` operator is
emitted for any statement of the form `lhs: InterfaceType =
module_container[idx]`. The same operator has to be used for both
`ModuleDict` and `ModuleList` because serialization does not preserve
the metadata that indicates whether a `Module` is a `ModuleDict` or
`ModuleList`.

**Testing**
This commit extends the existing unit tests for non-literal `ModuleDict`
indexing to test non-literal `ModuleList` indexing.

**Fixes**
This commit fixes #47496.

Differential Revision: [D26857597](https://our.internmc.facebook.com/intern/diff/D26857597)

Commit:
[JIT] Enable ModuleList non-literal indexing

**Summary**
This commit enables indexing into `ModuleList` using a non-literal
index if the LHS of the assignment statement of which the indexing is
the RHS is annotated with an interface type.

This feature already exists for `ModuleDict`, and this commit builds on
top of that implementation. A `prim::ModuleContainerIndex` operator is
emitted for any statement of the form `lhs: InterfaceType =
module_container[idx]`. The same operator has to be used for both
`ModuleDict` and `ModuleList` because serialization does not preserve
the metadata that indicates whether a `Module` is a `ModuleDict` or
`ModuleList`.

**Testing**
This commit extends the existing unit tests for non-literal `ModuleDict`
indexing to test non-literal `ModuleList` indexing.

**Fixes**
This commit fixes #47496.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -264,8 +264,36 @@ SugaredValuePtr ModuleValue::getitem(
-    return getSugaredDict(loc, m)->getModules()->getitem(
-        loc, m, idx, type_hint);
+    if (type_hint) {
+      // Check that all submodules comply with the type hint.
+      const auto& self_type = concreteType_->getJitType()->expect<ClassType>();
+      for (size_t i = 0; i < self_type->numAttributes(); ++i) {
+        const auto& attr_type = self_type->getAttribute(i);
+        if (attr_type->is_module()) {
+          std::stringstream ss;
+          if (!attr_type->isSubtypeOfExt(type_hint, &ss)) {
+            auto loc = self_->node()->sourceRange();
+            throw ErrorReport(loc)
+                << "Attribute " << self_type->getAttributeName(i)
+                << " is not of annotated type " << type_hint->annotation_str()
+                << ": " << ss.str();
+          }
+        }
+      }
```
Refactor this and the identical snippet below into a helper function.

gmagogsfm replied:
+1

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -623,11 +623,25 @@ RegisterOperators reg(
+     OperatorGenerator(
+         TORCH_SELECTIVE_SCHEMA(
+             "prim::ModuleContainerIndex.list(Any self, int ind) -> Any"),
+         [](Stack* stack) {
+           IValue ind = pop(stack);
+           IValue module_dict = pop(stack);
+           std::stringstream ss;
+           ss << ind.toInt();
+           push(stack, module_dict.toModule().attr(ss.str()));
+         },
+         aliasAnalysisFromSchema()),
+     // This operator is generated inside the compiler for indexing into
      // ModuleDict without a statically determinable key. Accordingly,
      // self must be a ModuleType and the output must be an InterfaceType.
      OperatorGenerator(
          TORCH_SELECTIVE_SCHEMA(
-             "prim::ModuleDictIndex(Any self, str ind) -> Any"),
+             "prim::ModuleContainerIndex.dict(Any self, str ind) -> Any"),
```
This is not BC, but I think I can make it BC by not changing the operator name, but making the `ind` parameter `Any` and using this operator for both `ModuleDict` and `ModuleList`. Thoughts?

gmagogsfm replied:
Based on my rough understanding, this doesn't affect serialized code. So it is not really BC-breaking I think. And AFAIK, we never made promise about operator backward-compatbility. Therefore, I think this is fine, let me know what you think.

eellison replied:
We only make operator-level BC promises about mobile & afaik this didnt work on mobile anyway, so this should be fine as @gmagogsfm said

gmagogsfm reviewed (approved):
Nice! Just have one comment that is probably best left for a different PR

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -220,11 +220,11 @@ class AttributePropagator {
         }
 
-        // Modules with prim::ModuleDictIndex cannot be frozen because they
+        // Modules with prim::ModuleContainerIndex cannot be frozen because they
```
We should try to get this to work. @gmagogsfm sometimes asks about what frontend team can do with perf. I think there's a lot to be said having frontend people work so that the new abstractions they introduce dont slow down user models. The alternative is that perf people stop making forward progress and begin working so that we dont regress.

SplitInfinity replied:
> We should try to get this to work.

What do you mean by this? We should try to get it to work without a custom operator?

eellison replied:
No, just add reasoning about the operator within freezing. Let's wait till we actually get an issue about this to do it.

SplitInfinity commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -264,8 +264,36 @@ SugaredValuePtr ModuleValue::getitem(
   if (concreteType_->getIterableModuleKind() == IterableModuleKind::LIST) {
-    return getSugaredDict(loc, m)->getModules()->getitem(
-        loc, m, idx, type_hint);
+    if (type_hint) {
```
Where is the annotation for `self.m` in your snippet?

gmagogsfm replied:
On line 17

Commit:
Update on "[JIT] Enable ModuleList non-literal indexing"


**Summary**
This commit enables indexing into `ModuleList` using a non-literal
index if the LHS of the assignment statement of which the indexing is
the RHS is annotated with an interface type.

This feature already exists for `ModuleDict`, and this commit builds on
top of that implementation. A `prim::ModuleContainerIndex` operator is
emitted for any statement of the form `lhs: InterfaceType =
module_container[idx]`. The same operator has to be used for both
`ModuleDict` and `ModuleList` because serialization does not preserve
the metadata that indicates whether a `Module` is a `ModuleDict` or
`ModuleList`.

**Testing**
This commit extends the existing unit tests for non-literal `ModuleDict`
indexing to test non-literal `ModuleList` indexing.

**Fixes**
This commit fixes #47496.

Differential Revision: [D26857597](https://our.internmc.facebook.com/intern/diff/D26857597)

[ghstack-poisoned]
